File: backend/code_java/readwordexample/readwordcode/src/main/java/com/code/countword/Implementation/CountWordImp.java
package com.code.countword.Implementation;

import com.code.countword.util.InfoWordCount;

import java.io.File;
import java.io.Serial;
import java.util.*;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.RecursiveTask;

import static java.util.stream.Collectors.toMap;

public class CountWordImp extends RecursiveTask<Long> {
    @Serial
    private static final long serialVersionUID = 1L;
    private final List<String> lines;
    private final File file;
    private final InfoWordCount infoWordCount;
    private final int numTask;
    private final int task;
    public CountWordImp(List<String> lines, File file, int task, int numTask, InfoWordCount infoWordCount){
        this.lines=lines;
        this.file=file;
        this.numTask=numTask;
        this.task=task;
        this.infoWordCount=infoWordCount;
    }

    @Override
    protected Long compute() {
        Map<String, Integer> wordCount = lines.stream().flatMap(line -> Arrays.stream(wordsIn(line))
                .filter(word -> word.length() >= this.infoWordCount.getLengthWord()))
                .filter(word -> word.length() > 0)
                .map(word -> new AbstractMap.SimpleEntry<>(word, 1))
                .collect(toMap(Map.Entry::getKey,Map.Entry::getValue, Integer::sum));
        this.infoWordCount.merge(wordCount, this.file,this.task,this.numTask);
        return 1L;
    }
    private String[] wordsIn(String line) {
        String[] s = line.trim().split("(\\s|\\p{Punct})+");
        return line.trim().split("(\\s|\\p{Punct})+");

    }

}

File: backend/code_scala/readwordexample/readwordcode/src/main/scala/com/readwordcode2/countword/implementation/CountWordImp.scala
package com.readwordcode2.countword.implementation

import com.readwordcode2.countword.contract.CountWord
import com.readwordcode2.countword.util.{AbstractExecutionContext, InfoWordCount}

import java.io.File
import scala.concurrent.Future

class CountWordImp(task:Int, numTask:Int, infoWordCount: InfoWordCount) extends AbstractExecutionContext with CountWord{
  override def countWordFork(lines: List[String], file: File):Future[Option[Long]]=Future{
    lines.flatMap(line=>wordsIn(line).filter(word=>word.length>=infoWordCount.getLengthWord))
      .filter(word=>word.nonEmpty).groupBy(value=>value).map(value=>value._1->value._2.size)}.map(result=>{
      infoWordCount.merge(result,file, task, numTask)
      Option(1L)
  })

  private def wordsIn(line: String): Array[String] =
    line.trim.split("(\\s|\\p{Punct})+")


}
